README.md
# NAME

CairoX::Sweet - Wraps Cairo for easier drawing

<div>
    <p>
    <img src="https://img.shields.io/badge/perl-5.10+-blue.svg" alt="Requires Perl 5.10+" />
    <a href="https://travis-ci.org/Csson/p5-CairoX-Sweet"><img src="https://api.travis-ci.org/Csson/p5-CairoX-Sweet.svg?branch=master" alt="Travis status" /></a>
    <a href="http://cpants.cpanauthors.org/release/CSSON/CairoX-Sweet-0.0200"><img src="http://badgedepot.code301.com/badge/kwalitee/CSSON/CairoX-Sweet/0.0200" alt="Distribution kwalitee" /></a>
    <a href="http://matrix.cpantesters.org/?dist=CairoX-Sweet%200.0200"><img src="http://badgedepot.code301.com/badge/cpantesters/CairoX-Sweet/0.0200" alt="CPAN Testers result" /></a>
    <img src="https://img.shields.io/badge/coverage-65.4%-red.svg" alt="coverage 65.4%" />
    </p>
</div>

# VERSION

Version 0.0200, released 2016-08-22.

# SYNOPSIS

    use CairoX::Sweet;

    my $c = CairoX::Sweet->new(500, 500, '#ffffff');
    my $path = CairoX::Sweet::Path->new(start => [35, 50], color => '#8855bb', width => 10, cap => 'round', join => 'round');

    $path->add_relative_line(qw/
        20 -5
        10 0
        30 -20
        -50 0
    /);
    $c->add_path($path, close => 1);

    $c->surface->write_to_png('a_short_line.png');

# DESCRIPTION

CairoX::Sweet is a wrapper around [Cairo](https://metacpan.org/pod/Cairo) which uses the great [cairo](http://www.cairographics.org/) graphics library.

For some use cases the standard api is a little verbose, the goal of this distribution is to reduce that.

# CONSTRUCTORS

## new($width, $height, %args)

    my $c = CairoX::Sweet->new(700, 500);
    my $c = CairoX::Sweet->new(700, 500, surface_format => 'argb32');

Both of these does this:

    my $surface = Cairo::ImageSurface->create('argb32', 700, 500);
    my $context = Cairo::Context->create($surface);

Use this constructor if you want to save the output as a png file later.

## new\_svg($width, $height, $filename, %args)

    my $c = CairoX::Sweet->new_svg(700, 500, 'output.svg');

Does this:

    Cairo::SvgSurface->create($filename, $width, $height);

Use this constructor if you want to save the output as an svg file later

# ATTRIBUTES

## width

## height

    my $c = CairoX::Sweet->new(700, 500);

Integers. Required. The width and height of the surface.

## background\_color

    my $c1 = CairoX::Sweet->new(..., background_color => '#83f9e2');
    my $c2 = CairoX::Sweet->new(..., background_color => [220, 75, 230]);
    my $c3 = CairoX::Sweet->new(..., background_color => [0.2, 0.4, 0.6]);

String or array reference of numbers. Optional. The starting background color.

You can pick one of three ways to give a color:

- As a hexadecimal string
- As an array reference giving a number each for red, green and blue (between 0 and 255).
- As an array reference in fractions between 0 and 1. This is the `cairo` style. **Note:** A color given as `[1, 1, 1]` is interpreted as (nearly) black, and not white as `cairo` would do.

## filename

    my $c = CairoX::Sweet->new_svg(..., filename => 'mysvg.svg');

String or a [Path::Tiny](https://metacpan.org/pod/Path::Tiny) path. Required for [new\_svg](#new_svg-width-height-filename-args).

# METHODS

## add\_path()

Adds a [CairoX::Sweet::Path](https://metacpan.org/pod/CairoX::Sweet::Path) to the `cairo` object.

    $c->add_path($path, close => 1);

`$path`

The [CairoX::Sweet::Path](https://metacpan.org/pod/CairoX::Sweet::Path) to add. Required.

`close =` Bool>

Boolean. Optional named parameter, defaults to `0`. If positive, will call `close_path|http://cairographics.org/manual/cairo-Paths.html#cairo-close-path`
on the `$path` (connects the two end points of the `$path`).

## add\_text()

Adds a string to the `cairo` object. Takes only named parameters.

    $c->add_text( text => "The text to add",
                  color => '#444444',
                  x => 37,
                  y => 115,
                  font_face => ['courier'],
                  font_size => 13
                  weight => 'bold',
    );

`text =` 'the text'>

String. The only required parameter. The text to add.

`x`

`y`

Integers. The position to start at.

`color`

A color. The color of the text. See ["background\_color"](#background_color) for more information on colors.

`font_face`

An array reference of font faces. There is no list of available font faces.

`font_size`

Number. The size of the text. Default is `10`.

`weight`

One of `normal` or `bold`.

`slant`

One of `normal`, `italic` or `oblique`.

## c()

Returns the [Cairo](https://metacpan.org/pod/Cairo) object itself.

    $c->c->set_source_rgb(0.4, 0.3, 0.2);
    $c->c->set_line_cap('round');
    $c->c->arc(150, 200, 35, 0, 3.1415);
    $c->c->stroke;

## surface()

Return the [Cairo surface](https://metacpan.org/pod/Cairo#Surfaces).

    $c->surface->write_to_png('a_short_line.png');

# STABILITY

Since this is an early release things might break, but hopefully not without good cause.

# PERFORMANCE

Using both [Moose](https://metacpan.org/pod/Moose) and [Moops](https://metacpan.org/pod/Moops) adds a little startup cost. If you need performance it is recommended to use [Cairo](https://metacpan.org/pod/Cairo) directly.

# SEE ALSO

- [CairoX::Sweet::Path](https://metacpan.org/pod/CairoX::Sweet::Path)
- [Cairo](https://metacpan.org/pod/Cairo)

# SOURCE

[https://github.com/Csson/p5-CairoX-Sweet](https://github.com/Csson/p5-CairoX-Sweet)

# HOMEPAGE

[https://metacpan.org/release/CairoX-Sweet](https://metacpan.org/release/CairoX-Sweet)

# AUTHOR

Erik Carlsson <info@code301.com>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2016 by Erik Carlsson.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for CairoX-Sweet

{{$NEXT}}
 [API Changes]

 [Bug Fixes]

 [Documentation]

 [Enhancements]

 [New Features]

 [Other Changes]

 [Requirements]

 [Stats & Dependency Changes]

 [Testing]

0.0200 2016-08-22T21:27:51Z UTC
 [Requirements]
 - Remove dependency on Moops.
 - Lower Perl compatibility to 5.10.

 [Stats & Dependency Changes]
 - 15 files changed, 668 insertions(+), 547 deletions(-)
 - (run req) + List::SomeUtils 0.53
 - (run req) + List::Util 1.45
 - (run req) + MooseX::AttributeShortcuts 0.028
 - (run req) + namespace::autoclean 0.22
 - (run req) ~ Moose 2.0000 --> 2.1803
 - (run req) ~ MooseX::StrictConstructor (any) --> 0.19
 - (run req) ~ Path::Tiny (any) --> 0.072
 - (run req) ~ Type::Tiny 1.000000 --> 1.000005
 - (run req) ~ Types::Path::Tiny (any) --> 0.005
 - (run req) ~ perl 5.014000 --> 5.010000
 - (run req) - List::AllUtils
 - (run req) - Moops
 - (test req) + IO::Handle (any)
 - (test req) + IPC::Open3 (any)
 - (conf req) + ExtUtils::MakeMaker (any)
 - (conf req) - Module::Build::Tiny
 - (dev req) + Badge::Depot 0.0103
 - (dev req) + Badge::Depot::Plugin::Coverage (any)
 - (dev req) + Badge::Depot::Plugin::Cpantesters (any)
 - (dev req) + Badge::Depot::Plugin::Kwalitee (any)
 - (dev req) + Badge::Depot::Plugin::Perl (any)
 - (dev req) + Badge::Depot::Plugin::Travis (any)
 - (dev req) + Dist::Zilla 5.000
 - (dev req) + Dist::Zilla::Plugin::Authority 1.009
 - (dev req) + Dist::Zilla::Plugin::BumpVersionAfterRelease::Transitional
   0.007
 - (dev req) + Dist::Zilla::Plugin::ChangeStats::Dependencies::Git 0.0100
 - (dev req) + Dist::Zilla::Plugin::Clean 0.07
 - (dev req) + Dist::Zilla::Plugin::DistIller::MetaGeneratedBy (any)
 - (dev req) + Dist::Zilla::Plugin::Git 2.036
 - (dev req) + Dist::Zilla::Plugin::Git::CheckFor::CorrectBranch 0.013
 - (dev req) + Dist::Zilla::Plugin::Git::Commit (any)
 - (dev req) + Dist::Zilla::Plugin::Git::Contributors 0.017
 - (dev req) + Dist::Zilla::Plugin::InstallRelease 0.008
 - (dev req) + Dist::Zilla::Plugin::MakeMaker (any)
 - (dev req) + Dist::Zilla::Plugin::MetaConfig (any)
 - (dev req) + Dist::Zilla::Plugin::NextRelease::Grouped 0.0200
 - (dev req) + Dist::Zilla::Plugin::PromptIfStale (any)
 - (dev req) + Dist::Zilla::Plugin::RewriteVersion::Transitional 0.007
 - (dev req) + Dist::Zilla::Plugin::Test::Compile 2.054
 - (dev req) + Dist::Zilla::Plugin::Test::EOF 0.0501
 - (dev req) + Dist::Zilla::Plugin::Test::EOL 0.18
 - (dev req) + Dist::Zilla::Plugin::Test::Kwalitee::Extra v0.2.1
 - (dev req) + Dist::Zilla::Plugin::Test::NoTabs 0.15
 - (dev req) + Dist::Zilla::Plugin::Test::Version 1.05
 - (dev req) + Pod::Weaver::Section::GenerateSection 1.01
 - (dev req) + Test::EOF (any)
 - (dev req) + Test::EOL (any)
 - (dev req) + Test::More 0.88
 - (dev req) + Test::NoTabs 1.4
 - (dev req) + Test::Version 1
 - (dev req) ~ Dist::Zilla::Plugin::CheckChangesHasContent (any) --> 0.008
 - (dev req) ~ Dist::Zilla::Plugin::CopyFilesFromBuild (any) --> 0.151680
 - (dev req) ~ Dist::Zilla::Plugin::GithubMeta (any) --> 0.54
 - (dev req) ~ Dist::Zilla::Plugin::MetaProvides::Package (any) -->
   2.003001
 - (dev req) ~ Dist::Zilla::Plugin::PodWeaver (any) --> 4.006
 - (dev req) ~ Dist::Zilla::Plugin::ReadmeAnyFromPod (any) --> 0.150250
 - (dev req) ~ Dist::Zilla::Plugin::RunExtraTests (any) --> 0.028
 - (dev req) ~ Dist::Zilla::Plugin::Test::ReportPrereqs (any) --> 0.021
 - (dev req) ~ Pod::Weaver::Section::Badges (any) --> 0.0401
 - (dev req) ~ Test::Warnings (any) --> 0.026
 - (dev req) - Dist::Zilla::Plugin::CheckExtraTests
 - (dev req) - Dist::Zilla::Plugin::ContributorsFromGit
 - (dev req) - Dist::Zilla::Plugin::MetaProvides::Class
 - (dev req) - Dist::Zilla::Plugin::MetaProvides::FromFile
 - (dev req) - Dist::Zilla::Plugin::ModuleBuildTiny
 - (dev req) - Dist::Zilla::Plugin::NextRelease
 - (dev req) - Dist::Zilla::Plugin::NextVersion::Semantic
 - (dev req) - Dist::Zilla::Plugin::OurPkgVersion
 - (dev req) - Dist::Zilla::Plugin::PodnameFromClassname
 - (dev req) - Dist::Zilla::Plugin::Prereqs::Plugins
 - (dev req) - Dist::Zilla::Plugin::PreviousVersion::Changelog
 - (dev req) - Dist::Zilla::Plugin::ReversionOnRelease
 - (dev req) - Pod::Weaver::Section::Homepage::DefaultCPAN
 - (dev req) - Pod::Weaver::Section::Source::DefaultGitHub
 - (dev sug) + Dist::Iller 0.1408
 - (dev sug) + Dist::Iller::Config::Author::CSSON 0.0312

0.0108 2015-03-21T19:36:20Z CET
 [Documentation]
 - Added badges (via Pod::Weaver::Section::Badges).

0.0107 2015-03-19T16:43:43Z CET
 [Meta]
 - iller.yaml is now used to build dist.ini and weaver.ini
 - Various cleanup:
 - Prereqs are now set in dist.ini instead of cpanfile.
 - Changed some # PODNAME to # PODCLASSNAME.
 - Added strict and warnings to a couple of classes (for kwalitee).
 - Moved a couple of # VERSION to inside packages/classes.

 [Requirements]
 - Fixed missing prereqs.

0.0106 2015-01-19T21:10:31Z CET
 [Bug Fixes]
 - Had more failures for truth, so added 1; to all Moops classes.

0.0105 2015-01-17T12:56:27Z CET
 [Bug Fixes]
 - Added truth to Color (had some test failures).

 [Enhancements]
 - Require Cairo in test.

 [Meta]
 - Require latest Moops (0.034).
 - New test prereq Test::Requires.

0.0104 2015-01-16T11:49:12Z CET
 [Bug Fixes]
 - Removed a faulty use statement.
 - Now requires the right Perl version (5.14)

 [Meta]
 - Now adds all classes (and type library) to provides.

0.0103 2015-01-11T12:40:38Z CET
 [Documentation]
 - Remove empty abstracts

0.0102 2015-01-11T12:24:34Z CET
 [Meta]
 - Set podname, cleanup abstracts
 - Cleanup prereqs

0.0101 2015-01-11T00:57:30Z CET
 [Meta]
 - Added prereqs.

0.0100 2015-01-11T00:49:25Z CET
 - Initial release


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.007.
use strict;
use warnings;

use 5.010000;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Wraps Cairo for easier drawing",
  "AUTHOR" => "Erik Carlsson <info\@code301.com>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "CairoX-Sweet",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.010000",
  "NAME" => "CairoX::Sweet",
  "PREREQ_PM" => {
    "Cairo" => "1.000",
    "Eponymous::Hash" => 0,
    "List::SomeUtils" => "0.53",
    "List::Util" => "1.45",
    "Moose" => "2.1803",
    "MooseX::AttributeShortcuts" => "0.028",
    "MooseX::StrictConstructor" => "0.19",
    "Path::Tiny" => "0.072",
    "Type::Tiny" => "1.000005",
    "Types::Path::Tiny" => "0.005",
    "namespace::autoclean" => "0.22"
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::More" => "0.96",
    "Test::Requires" => "0.08"
  },
  "VERSION" => "0.0201",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Cairo" => "1.000",
  "Eponymous::Hash" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "List::SomeUtils" => "0.53",
  "List::Util" => "1.45",
  "Moose" => "2.1803",
  "MooseX::AttributeShortcuts" => "0.028",
  "MooseX::StrictConstructor" => "0.19",
  "Path::Tiny" => "0.072",
  "Test::More" => "0.96",
  "Test::Requires" => "0.08",
  "Type::Tiny" => "1.000005",
  "Types::Path::Tiny" => "0.005",
  "namespace::autoclean" => "0.22"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
